// @lightningjs/blits/src/components/Circle.js
/*
 * Copyright 2023 Comcast Cable Communications Management, LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import Component from '../component.js'

export default () =>
  Component('Circle', {
    /* eslint-disable no-unused-vars */
    code: {
      render: function anonymous(parent, component, context, components, effect, getRaw, Log) {
        const elms = []
        const elementConfigs = []
        const forloops = []
        const props = []
        const created = []
        const effects = {}
        const skips = []
        let componentType
        let rootComponent = component
        let propData
        let slotComponent
        let inSlot = false
        let slotChildCounter = 0
        let cmps = []
        elementConfigs[0] = {}

        elms[0] = this.element(
          { parent: parent || 'root' },
          inSlot === true ? slotComponent : component
        )

        elementConfigs[0]['color'] = component.color
        elementConfigs[0]['w'] = component.size
        elementConfigs[0]['h'] = component.size
        elementConfigs[0]['effects'] = [component.shader('radius', { radius: component.radius })]
        elms[0].populate(elementConfigs[0])

        if (inSlot === true) {
          slotChildCounter -= 1
        }

        return {
          elms,
          cleanup: () => {
            rootComponent = null
            propData = null
            slotComponent = null
            parent = null
            component = null
            cmps.length = 0
            elms.length = 0
            components.length = 0
            elementConfigs.length = 0
            forloops.length = 0
            props.length = 0
            skips.length = 0
          },
        }
      },
      effects: [
        function anonymous(component, elms, context, components, rootComponent, effect) {
          elms[0].set('color', component.color)
        },
        function anonymous(component, elms, context, components, rootComponent, effect) {
          elms[0].set('w', component.size)
        },
        function anonymous(component, elms, context, components, rootComponent, effect) {
          elms[0].set('h', component.size)
        },
        function anonymous(component, elms, context, components, rootComponent, effect) {
          elms[0].set('effects', [component.shader('radius', { radius: component.radius })])
        },
      ],
      context: {},
    },
    props: [{ key: 'size', default: 40 }, 'color'],
    computed: {
      radius() {
        return this.size / 2
      },
    },
  })
