// @lightningjs/blits/src/components/RouterView.js
/*
 * Copyright 2023 Comcast Cable Communications Management, LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import Component from '../component.js'
import Router from '../router/router.js'
import symbols from '../lib/symbols.js'
import Focus from '../focus.js'

let hashchangeHandler = null

export default () =>
  Component('RouterView', {
    /* eslint-disable no-unused-vars */
    code: {
      render: function anonymous(parent, component, context, components, effect, getRaw, Log) {
        const elms = []
        const elementConfigs = []
        const forloops = []
        const props = []
        const created = []
        const effects = {}
        const skips = []
        let componentType
        let rootComponent = component
        let propData
        let slotComponent
        let inSlot = false
        let slotChildCounter = 0
        let cmps = []
        elementConfigs[0] = {}

        elms[0] = this.element(
          { parent: parent || 'root' },
          inSlot === true ? slotComponent : component
        )

        elementConfigs[0]['w'] = parent.node.width * (100 / 100)
        elementConfigs[0]['height'] = parent.node.height * (100 / 100)
        elms[0].populate(elementConfigs[0])

        if (inSlot === true) {
          slotChildCounter -= 1
        }

        return {
          elms,
          cleanup: () => {
            rootComponent = null
            propData = null
            slotComponent = null
            parent = null
            component = null
            cmps.length = 0
            elms.length = 0
            components.length = 0
            elementConfigs.length = 0
            forloops.length = 0
            props.length = 0
            skips.length = 0
          },
        }
      },
      effects: [],
      context: {},
    },
    state() {
      return {
        activeView: null,
      }
    },
    hooks: {
      async ready() {
        if (this.parent[symbols.routerHooks] && this.parent[symbols.routerHooks].init) {
          await this.parent[symbols.routerHooks].init.apply(this.parent)
        }
        hashchangeHandler = () => Router.navigate.apply(this)
        Router.navigate.apply(this)
        window.addEventListener('hashchange', hashchangeHandler)
      },
      destroy() {
        window.removeEventListener('hashchange', hashchangeHandler, false)
      },
      focus() {
        if (this.activeView && Focus.get() === this) {
          this.activeView.$focus()
        }
      },
    },
    input: {
      back(e) {
        const navigating = Router.back.call(this)
        if (navigating === false) {
          this.parent.$focus(e)
        }
      },
    },
  })
